### PR TITLE
Added D2 Regression Metrics

### PR DESCRIPTION
**Pull request by anmolsjoshi**
Related to #307 

Description:
* Canberra Metric
* Fractional Absolute Error
* Wave Hedges Distance
* Geometric Mean Absolute Error
* ~~Geometric Mean Relative Absolute Error~~

Check list:
* [x] New tests are added (if a new feature is modified)
* [x] New doc strings: text and/or example code are in RST format
* [x] Documentation is updated (if required)

### CODE REVIEW

**Comment by anmolsjoshi** [comment]
Should have the tests completed in a few hours!

**Comment by anmolsjoshi** [comment]
cc @vfdev-5 @zasdfgbnm

File: ignite/contrib/metrics/regression/geometric_mean_absolute_error.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        y_pred, y = output
+        errors = torch.abs(y.view_as(y_pred) - y_pred)
+        self._product_of_errors = self._product_of_errors * torch.prod(errors).item()
+        self._num_examples += y.shape[0]
+
+    def compute(self):
+        if self._num_examples == 0:
+            raise NotComputableError('GeometricMeanAbsoluteError must have at'
+                                     'least one example before it can be computed.')
+        return np.power(self._product_of_errors, 1.0 / self._num_examples)
```
Comment (zasdfgbnm) [inline ignite/contrib/metrics/regression/geometric_mean_absolute_error.py]: I suggest to replace `np.power(a, b)` with `a ** b`, since numpy is not a dependency of ignite.

File: ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    def update(self, output):
+        y_pred, y = output
+        errors = torch.abs(y.view_as(y_pred) - y_pred)
+        # previous sum is required to calculate new average of ground truth
+        prev_sum_of_gt = self._mean_of_actuals * self._num_examples
+        self._num_examples += y.shape[0]
+        # once num_examples is updated, new average is (prev_sum + y.sum())/num_examples
+        self._mean_of_actuals = (prev_sum_of_gt + y.sum().item()) / self._num_examples
+        relative_error = errors / torch.abs(y - self._mean_of_actuals)
+        self._product_of_errors = self._product_of_errors * torch.prod(relative_error).item()
```
Comment (zasdfgbnm) [inline ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py]: I don't think this metric is possible to implement without saving all the history. Since the A bar is unknown until the last sample come in, it is impossible to compute the denominator with the correct value of A bar. The implementation now looks not correct to me because in `_product_of_errors` is computed using old A bar, rather than the correct value of A bar.

File: ignite/contrib/metrics/regression/geometric_mean_absolute_error.py (hunk 1 above)
Comment (anmolsjoshi) [reply ignite/contrib/metrics/regression/geometric_mean_absolute_error.py]: Agree will incorporate tonight.

File: ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py (hunk 2 above)
Comment (anmolsjoshi) [reply ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py]: Ok! I’ll remove this from this PR but correct the implementation separately. We can add it once we decided how to handle metics that require the entire history of the training.

**Comment by anmolsjoshi** [comment]
@vfdev-5 @zasdfgbnm I have incorporated comments

- removed GMRAE implementation pending further discussion about recording the entire history of training
- changed calculation of power to remove numpy

let me know if there are any issues

**Comment by anmolsjoshi** [comment]
cc @vfdev-5

File: ignite/contrib/metrics/regression/canberra_metric.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+from __future__ import division
+import torch
+from ignite.metrics.metric import Metric
+
+
+class CanberraMetric(Metric):
+    r"""
+    Calculates the Canberra Metric.
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/canberra_metric.py]: Could you please unify the docstring like for other implementations, e.g
```
    Calculates the Manhattan Distance:
    :math:`\text{MD} = \sum_{j=1}^n (A_j - P_j)`,
    where :math:`A_j` is the ground truth and :math:`P_j` is the predicted value.
    More details can be found in `Botchkarev 2018`__.
    - `update` must receive output of the form `(y_pred, y)`.
    - `y` and `y_pred` must be of same shape.
    __ https://arxiv.org/abs/1809.03006

```

File: ignite/contrib/metrics/regression/wave_hedges_distance.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    - `y` and `y_pred` must be of same shape.
+    __ https://arxiv.org/abs/1809.03006
+    """
+
+    def reset(self):
+        self._sum_of_errors = 0.0
+
+    def update(self, output):
+        y_pred, y = output
+        errors = torch.abs(y.view_as(y_pred) - y_pred) / torch.max(y_pred, y.view_as(y_pred))
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/wave_hedges_distance.py]: I have a doubt on this one, I'm not sure that $max(A_j, P_j)$ = `torch.max(y_pred, y.view_as(y_pred))`. Need to check this from my side

**Comment by vfdev-5** [comment]
@anmolsjoshi thanks for pinging ! I left some comments. Otherwise looks good.

File: ignite/contrib/metrics/regression/canberra_metric.py (hunk 3 above)
Comment (anmolsjoshi) [reply ignite/contrib/metrics/regression/canberra_metric.py]: Done!

File: ignite/contrib/metrics/regression/wave_hedges_distance.py (hunk 4 above)
Comment (anmolsjoshi) [reply ignite/contrib/metrics/regression/wave_hedges_distance.py]: @vfdev-5 see code below

```python
y_pred = torch.rand(10)
y = torch.rand(10)
maxes = torch.max(y_pred, y)

print (y_pred)
# tensor([0.1823, 0.0923, 0.1578, 0.9375, 0.4769, 0.0928, 0.7895, 0.2848, 0.4338, 0.2930])
print (y)
# tensor([0.3095, 0.8486, 0.2325, 0.3083, 0.6410, 0.8848, 0.5019, 0.0213, 0.1661, 0.2154])
print (maxes)
# tensor([0.3095, 0.8486, 0.2325, 0.9375, 0.6410, 0.8848, 0.7895, 0.2848, 0.4338, 0.2930])
```

**Comment by anmolsjoshi** [comment]
@vfdev-5 thanks for the review. responded to comments.

File: ignite/contrib/metrics/regression/wave_hedges_distance.py (hunk 4 above)
Comment (vfdev-5) [reply ignite/contrib/metrics/regression/wave_hedges_distance.py]: You are right, `torch.max` is element-wise max if two tensors are compared.

**Comment by vfdev-5** [comment]
I wonder also whether we should check input data shape ?

**Comment by anmolsjoshi** [comment]
yes, I think this is something required in all the regression metrics, currently none of them employ it. It can be as simple as:

```python
if y_pred.shape == y.shape:
    raise ValueError('......')
```

File: ignite/contrib/metrics/regression/geometric_mean_absolute_error.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import torch
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric
+
+
+class GeometricMeanAbsoluteError(Metric):
+    r"""
+    Calculates the Geometric Mean Absolute Error.
+
+    :math:`\text{GMRAE} = \sqrt[n]{\prod_{j=1}^{\n}{|A_j - P_j|}`
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/geometric_mean_absolute_error.py]:
```suggestion
    :math:`\text{GMRAE} = \sqrt[n]{\prod_{j=1}^{n}{|A_j - P_j|}}`
```

**Comment by vfdev-5** [comment]
That's true that prediction and true values should have the same shape, but shouldn't they be just vectors : `(N, )` ?

File: ignite/contrib/metrics/regression/geometric_mean_absolute_error.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    """
+
+    def reset(self):
+        self._product_of_errors = 1.0
+        self._num_examples = 0
+
+    def update(self, output):
+        y_pred, y = output
+        errors = torch.abs(y.view_as(y_pred) - y_pred)
+        self._product_of_errors = self._product_of_errors * torch.prod(errors).item()
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/geometric_mean_absolute_error.py]:
```suggestion
        self._product_of_errors *= torch.prod(errors).item()
```

Comment (vfdev-5) [inline ignite/contrib/metrics/regression/geometric_mean_absolute_error.py]: I think we should implement this differently (e.g. with log) in order to underflow when multiply error values ~1e-1, 1e-2.

Comment (anmolsjoshi) [reply ignite/contrib/metrics/regression/geometric_mean_absolute_error.py]: Agree. I have a working local implementation I can push

**Comment by anmolsjoshi** [comment]
So would the shape check confirm that y and y_pred are the same shape and (N, ) or (N, 1) ?

File: ignite/contrib/metrics/regression/geometric_mean_absolute_error.py (hunk 6 above)
Comment (anmolsjoshi) [reply ignite/contrib/metrics/regression/geometric_mean_absolute_error.py]: Added implementation using log.

**Comment by anmolsjoshi** [comment]
@vfdev-5 I updated the implementation of GMAE using log and exp. I left the tests the same to see if I got the same answers and I did. Let me know if you want me to update the test to incorporate the current implementation. 

Also what are your thoughts on creating a `BaseRegressionMetric` class that include a shape check as we discussed above. All regression modules can inherit this. Shape check, checks if y_pred and y are (N) or (N, 1), it squeezes both to (N,).

**Comment by vfdev-5** [comment]
> Also what are your thoughts on creating a BaseRegressionMetric class that include a shape check as we discussed above. All regression modules can inherit this. Shape check, checks if y_pred and y are (N) or (N, 1), it squeezes both to (N,).

@anmolsjoshi seems good.

**Comment by vfdev-5** [comment]
Curiously exp-log implementation of GMAE works without any explosions on zero error:
```python
y_true = torch.Tensor([0.0, 0.0, 0.0, 0.0])
y_pred = torch.Tensor([0.01, 0.0, 0.0001, 0.2])

gmae.update((y_pred, y_true))
gmae._sum_of_errors, gmae.compute()
> tensor(-inf), 0.0
```

**Comment by vfdev-5** [comment]
@anmolsjoshi @jasonkriss please review 

@jasonkriss merge if ok

**Comment by anmolsjoshi** [comment]
@vfdev-5 LGTM! Thanks for the last few commits!

**Comment by jasonkriss** [comment]
LGTM! Nice work guys!
